### DISCUSSION TITLE
Error while trying to run YOLOv8 training on a custom dataset

### DISCUSSION THREAD

**Discussion by Timotej979** [ultralytics/ultralytics#8748] [Q&A] [answered]
Hello,

so i get the following error trying to run the YOLOv8 training on a custom dataset :thinking:  ([Link to repo](https://github.com/Timotej979/U-Net-Water-Land-Segmentation)   __/container/detection__ folder and __/container/dataset/contour-det__ folder). 

Here is the error (I have found a similar error type for the YOLOv5 model, but not for the v8 version):
```
Traceback (most recent call last):
unet_yolo_container-1  |   File "/app/container/detection/run.py", line 316, in <module>
unet_yolo_container-1  |     model_controler.train_val_contour()
unet_yolo_container-1  |   File "/app/container/detection/run.py", line 97, in train_val_contour
unet_yolo_container-1  |     model.train(data=contour_ds, epochs=self.opt.epochs, batch=-1, device=self.device)
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/engine/model.py", line 371, in train
unet_yolo_container-1  |     self.trainer.train()
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/engine/trainer.py", line 191, in train
unet_yolo_container-1  |     self._do_train(world_size)
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/engine/trainer.py", line 362, in _do_train
unet_yolo_container-1  |     self.metrics, self.fitness = self.validate()
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/engine/trainer.py", line 462, in validate
unet_yolo_container-1  |     metrics = self.validator(self)
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
unet_yolo_container-1  |     return func(*args, **kwargs)
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/engine/validator.py", line 163, in __call__
unet_yolo_container-1  |     self.loss += trainer.criterion(preds, batch)[1]
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/v8/detect/train.py", line 77, in criterion
unet_yolo_container-1  |     return self.compute_loss(preds, batch)
unet_yolo_container-1  |   File "/usr/local/lib/python3.8/site-packages/ultralytics/yolo/v8/detect/train.py", line 194, in __call__
unet_yolo_container-1  |     target_scores_sum = max(target_scores.sum(), 1)
unet_yolo_container-1  | RuntimeError: Inplace update to inference tensor outside InferenceMode is not allowed.You can make a clone to get a normal tensor before doing inplace update.See https://github.com/pytorch/rfcs/pull/17 for more details.
```

Here are the requirements used in the Dockerfile:
```
# Python requirements file

# Segmentation model dependencies
torch==1.9.0
torchvision==0.10.0
torchaudio==0.9.0
torchmetrics
numpy
opencv-python
tqdm
wandb
albumentations
matplotlib

# YOLOv8 dependencies
ultralytics

# AutoDistill dependencies
transformers
autodistill
autodistill-grounded-sam
autodistill-yolov8
roboflow
supervision==0.9.0
```

This is my config.yaml:
```yaml
train: /app/container/dataset/contour-det/images/train
val: /app/container/dataset/contour-det/images/val
test: /app/container/dataset/contour-det/images/test
nc: 1
names: ['obstacle']
```

and this is the code i am trying to run:
```python
def train_val_contour(self):
        # Initialize a new run
        self.initialize_new_run("YOLOv8-contour-model-train")

        # Define the dataset configuration
        contour_ds = self.opt.dataset_root + '/contour-det/config.yaml'
        # Load the YOLOv8 model depending on the size
        if self.opt.model_size == 'n':
            model = YOLO('/app/container/detection/weights/yolov8n.pt')
        elif self.opt.model_size == 's':
            model = YOLO('/app/container/detection/weights/yolov8s.pt')
        elif self.opt.model_size == 'm':
            model = YOLO('/app/container/detection/weights/yolov8m.pt')
        elif self.opt.model_size == 'l':
            model = YOLO('/app/container/detection/weights/yolov8l.pt')
        elif self.opt.model_size == 'x':
            model = YOLO('/app/container/detection/weights/yolov8x.pt')
        else:
            print("Invalid model size")
            exit(1)

        # Train the model
        if self.device == 'cpu':
            results = model.train(data=contour_ds, epochs=self.opt.epochs, batch=-1, device=self.device)
        else:
            results = model.train(data=contour_ds, epochs=self.opt.epochs, batch=-1, device=0)
        
        print(results)

        # Stop wandb logging
        wandb.finish()

        # Initialize a new run
        self.initialize_new_run("YOLOv8-contour-model-val")

        # Validate the model
        if self.device == 'cpu':
            metrics = model.val(data=contour_ds, batch=-1, device=self.device)
        else:
            metrics = model.val(data=contour_ds, batch=-1, device='0')

        print(metrics)

        # Stop wandb logging 
        wandb.finish()
```

Help is much appreciated :pray:  since it is quite an obscure error, maybe something to do with the environment setup or maybe I have to set the imgsz parameter if I am using images of custom sizes, since the code is meant to mostly work out of the box? :smiley:

**Comment by glenn-jocher** [comment, marked as answer]
Hello! 👋 It looks like you're encountering an issue related to PyTorch's InferenceMode with your custom dataset training. This error often arises due to in-place operations on tensors that require gradients, which are not allowed outside of PyTorch's InferenceMode.

Given the error and your setup, I'd recommend two things to try:

1. **Update PyTorch and Dependencies**: Your current setup uses PyTorch `1.9.0`, which is relatively older. There have been significant updates and bug fixes since then. Try updating PyTorch and its related libraries (`torchvision`, `torchaudio`) to their latest versions. This might resolve compatibility issues with the latest `ultralytics` package.

    ```bash
    pip install torch torchvision torchaudio --upgrade
    ```

2. **Review In-Place Operations**: If the issue persists, it might be related to specific in-place operations in your code or within the YOLOv8 implementation you're using. While it's more challenging to debug without seeing the full codebase, ensure that any tensor modifications are not done in-place on tensors that are part of the computation graph. For example, use `tensor = tensor.clone()` before in-place operations like `tensor += 1`.

Regarding your question about `imgsz`, it's always a good idea to specify `imgsz` if your images are of custom sizes, especially if they deviate significantly from the default size. This ensures the model processes them correctly during training and validation.

If these suggestions don't resolve the issue, it might be helpful to reach out with more details or consider opening an issue in the Ultralytics YOLOv8 GitHub repository for more specific guidance. The community and developers there can provide more targeted support. 😊

Hope this helps, and best of luck with your training!

**Comment by Timotej979** [comment]
Hello,

thanks for the answer and help 😊 , as suggested i switched the PyTorch version to 2.2.0 and the training succeeded. :smile:
I have also specified the image size to ensure the correct processing of images during training and validation.

**Reply by glenn-jocher** [reply]
That's fantastic news! 😊 I'm glad to hear that updating PyTorch to version 2.2.0 and specifying the image size resolved the issue for you. If you have any more questions or need further assistance as you continue working with YOLOv8, feel free to reach out. Happy training! 🚀

**Reply by Manjutiwari** [reply]
---------------------------------------------------------------------------



















































My model was running perfectly till last few days...all of sudden started encountaring the troublr
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[17], line 1
----> 1 from ultralytics import YOLO

ModuleNotFoundError: No module named 'ultralytics'

Please help in debugging as my deadline of submission is near

**Reply by glenn-jocher** [reply]
@Manjutiwari it sounds like there might be an issue with your Python environment. A quick solution could be to ensure that the `ultralytics` package is correctly installed. Try reinstalling it using this command:

```bash
pip install ultralytics
```

If you're working in a virtual environment, make sure it's activated before installing. Also, double-check that you're using the correct Python interpreter where `ultralytics` is installed. Hopefully, this fixes the issue! 🛠️ If not, feel free to provide more details.

**Reply by Manjutiwari** [reply]
Thanks, the problem seems to be with the virtual environment and is resolved.

**Reply by glenn-jocher** [reply]
Great to hear the problem is resolved! 🎉 If you have any more questions or run into other issues, feel free to reach out. Happy coding with YOLOv8! 😊

**Comment by Manjutiwari** [comment]
I am trying to detect grass in rice using yolov8m.pt. My training dataset
contains 17000 bounding boxes in 42 images (image size 8000*6000 mp) in
three classes: rice, grass, and background. I am trying to tune it but am
not able to get an accurate result. Can you help or suggest how to improve?
so far, I am using the training parameters, like
 model.train(data = "data_custom.yaml", batch=2, imgsz=1600, rect=True,
epochs=100,lr0=0.001, optimizer='Adam', workers=1)

**Reply by pderrenger** [reply]
@Manjutiwari hey there! 🌾 It looks like you're working on an intriguing project with YOLOv8 for detecting grass in rice fields. Given the size of your images (8000x6000) and the dense population of bounding boxes per image, adjusting the following might help improve your model's accuracy:

1. **Image Size and Tiling:** Your `imgsz=1600` with `rect=True` is a good start, but considering the high resolution of your images, you can potentially increase the `imgsz` even further if your hardware allows. Alternatively, consider segmenting your large images into smaller tiles, ensuring each tile contains sufficient instances of your classes for effective training.

2. **Data Augmentation:** Use various data augmentation techniques to expose your model to a wider variety of training scenarios. This can include random flips, rotations, color space augmentations, etc. It helps the model generalize better.

3. **Learning Rate Scheduler:** Experiment with different learning rate schedules. A lower starting learning rate (`lr0=0.001`) is good, but you might benefit from a scheduler that adjusts the rate based on validation metrics improvement or epochs.

4. **Optimizer:** While you're using `Adam`, it's worth experimenting with `SGD` with a momentum as it sometimes outperforms Adam for such large-scale image datasets.

5. **Focus on Hard Examples:** Given your scenario with potentially many small objects (grass) in a large image, paying more focus on hard examples during training might help. This technique involves giving more weight to losses from hard-to-detect objects.

Remember, the key to tuning is iteration and evaluation. After adjusting parameters, closely monitor your model's performance on a validation set to see if the changes yield positive results.

Keep up the great work, and don't hesitate to reach out if you have any more questions or need further assistance!
